cpp/src/traversal/opg/bfs_comms.cuh

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <raft/handle.hpp>
+#include <graph.hpp>
+
+
+namespace detail {
+
+namespace opg {
+
+constexpr inline size_t
+divup(const size_t& numerator, const size_t& denominator) {
+  return (numerator + denominator - 1)/denominator;
+}
+
+size_t number_of_words(size_t number_of_bits)
+{
+  size_t numerator = number_of_bits;
+  size_t denominator = BitsPWrd<unsigned>;
+  return (numerator > denominator) ?
+    1 + divup(numerator - denominator, denominator) :
+    (denominator > 0);
+}
+
+__global__
+void reduce_bitwise_or(
+    unsigned** frontiers,
+    size_t frontier_count,
+    int my_rank,
+    size_t word_count) {
+
+  unsigned word = 0;
+  size_t tid = threadIdx.x + (blockIdx.x * blockDim.x);
+  if (tid < word_count) {
+    for (int i = 0; i < frontier_count; ++i) {
+      word = word | frontiers[i][tid];
+    }
+    frontiers[my_rank][tid] = word;
+  }
+}
+
+template <typename VT, typename ET, typename WT>
```

Comment:
The primary reason for having bfs communicator right now is because we do not have bitwise or all reduce in raft. This class allows me to abstract out the resource allocation and calls for this specific use case.

For experimentation stage with custom nccl, I plan to switch the internals of this class so that I can easily test the performance difference without a lot of code changes.

Once a particular implementation is selected, I intend to remove this class.

Reply:
In that case, better add a comment saying this code is for temporary testing and will not be merged. Then, I will worry less about code bifurcation. We'd better avoid code duplication/bifurcation as this makes our codebase very difficult to maintain and understand.